Approved/enhanced-array-literals.md

```diff
@@ -0,0 +1,149 @@
+
+This assumption is also invalid for uninhabited types such as `Void`.
+If Q# requires that every type has a default value, then uninhabited types are not possible to express properly.
+This can even cause subtle bugs in the soundness of a type system, as demonstrated by [this bug in Java's generics](https://hackernoon.com/java-is-unsound-28c84cb2b3f) that happened because of the existence of a value (specifically `null`) for a type that should have been uninhabited.
+
+## Enhancing array literals
+
+With the removal of default-initialized array constructors, alternatives are needed to create empty arrays and arrays with a repeated initial value.
+The alternative for `new T[0]` is `[]`.
+The alternative for `new T[n]` where `n > 0` is `[x, size = n]`, where `x` is the initial value of each item.
+
+# Description
+
+## Current Status
+
+Q# currently supports this syntax:
+
+```qsharp
+// Create an empty array of type Int[].
+let empty = new Int[0];
+
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = new Bool[10];
+```
+
+## Proposed Modification
+
+The syntax above will be deprecated and removed.
+The replacement for each case is different.
+
+### Empty arrays
+
+The empty array literal, `[]`, will be supported.
+Unlike `new T[0]`, this literal does not specify the type of the array to be created.
+Its type will be inferred based on its first usage.
+
+```qsharp
+function SumI(xs : Int[]) : Int {
+    return Fold(PlusI, 0, xs);
+}
+
+function SumD(xs : Double[]) : Int {
+    return Fold(PlusD, 0.0, xs);
+}
+
+function Example() : Unit {
+    // The type of [] is inferred to be Int[] based on its position as the argument to SumI.
+    let x = SumI([]);
+
+    // The type of empty1 is inferred to be Int[] based on its position as the argument to SumI.
+    let empty1 = [];
+    let y = SumI(empty1);
+
+    // The following line contains a compilation error, because empty1 is of type Int[], not Double[].
+    let z = SumD(empty1);
+
+    // Since empty2 is not used, its type cannot be inferred.
+    // This is a compilation error.
+    let empty2 = [];
+
+    // While empty3 is used, its usage does not provide any information about its concrete type.
+    // This is a compilation error.
+    let empty3 = [];
+    Message($"Empty array: {empty3}");
+
+    // In cases where the type of [] cannot be inferred, Microsoft.Quantum.Arrays.EmptyArray can be used instead.
+    // The following code is valid.
+    let empty4 = EmptyArray<Int>();
+    Message($"Empty array: {empty4}");
+}
+```
+
+### Repeated constant arrays
+
+The semantics of `[x, size = n]` are identical to `new T[n]`, except that the initial value is now given by `x`:
+
+1. `n` may be an integer literal or a variable of integer type.
+2. If `n` is negative, a runtime error occurs.
+
+```qsharp
+// Create an array of length 10 of type Bool[] where every value is false.
+let bools = [false, size = 10];
```

Comment:
I'm still concerned about the inconsistent terminology: to create an array of *length* n, use the *size* option...

Reply:
Well, when we introduce multi-dimensional arrays it would make sense to introduce a Size function that can be used also for 1D arrays. I think using Length for multidimensional arrays really doesn't work, and it would seem odd to not support Size for the 1D case. @cgranade Does supporting a Size function for arrays of any dimension make sense to you and if so, can we introduce support for it (for the 1D case) before we introduce multidimensional arrays? I would suggest to keep Length, which would mean that both Length and Size are valid and return the same for the 1D case.